doc/case-studies/timelock/README.md

```diff
@@ -0,0 +1,242 @@
+2. Side effects introduced during system evolution: same as above. Additionally, if an effect monitor is configured to revert in case of unexpected effects taking place, the developers will be forced to keep the spec in sync with the implementation.
+3. Inconsistent / spread-out specs: An effect monitor may describe all effects that may happen wrt. a particular state component in one place. As this monitor will be triggered when any of the methods that modify the state is executed, this also brings us considerable savings in terms of spec size/complexity, as similar effects can be grouped together.
+4. Unrestricted access from other contracts/methods: as in 1. and 2., it doesn't matter again where the modification comes from: if the state component we monitor is being changed, the monitor will detect it.
+
+
+## A use case: Timelock monitor specs
+
+Now it's time to look at the concrete example, the [Timelock contract][]. Go ahead, and compare:
+
+- Direct monitor specs in the [SCF 24 submission example][]
+- [Timelock monitor spec][] written in the style outlined here
+
+Let's go over the spec together here. (please note that this spec is not even a completely valid TLA+ yet, as e.g. many definitions need to be filled in from the contract code; this is tbd.)
+
+### Direct monitor for the `deposit` method
+
+The [deposit method](https://github.com/stellar/soroban-examples/blob/186443aab0bf8b7c2673428c38708bf38cb772ab/timelock/src/lib.rs#L57-L91) allows a user to send certain amount of token to this contracts, and to specify up to 10 claimants any of whom should be able to claim the funds when the specified timing condition fulfills. `MustFail` predicates in the spec closely follow the assertions in the code. Notice though that `MustFail_deposit_NotEnoughBalance` is not an assertion in the code, but a precondition which arises from usage of an external token contract. Nevertheless, the monitor system will force us to be complete (otherwise it will warn us about an unexpectedly reverted transaction).
+
+
+```tla 
+MustFail_deposit_TooManyClaimants == 
+    claimants.len > 10
+
+MustFail_deposit_Unauthorized == 
+    ~authorized(from)
+
+\* This property would not be originally conceived 
+\* if looking solely at the timelock contract source
+MustFail_deposit_NotEnoughBalance == 
+    token.balance(from) < amount
+
+\* Balance is externally relevant; initialized flag is not
+\* What matters for users is that balance is not overwritten
+MustFail_deposit_AlreadyInitialized == 
+    exists(balance)
+
+\* The above failure conditions exhaust the precondition of deposit method
+\* The default success condition is not needed, but we may provide it
+MustSucceed_deposit_Default = TRUE
+
+MustAchieve_deposit_BalanceRecordCreated ==
+    exists(balance')
+
+MustAchieve_deposit_BalanceRecordCorrect ==
+    /\ balance'.token = token
+    /\ balance'.amount = amount
+    /\ balance'.time_bound = time_bound
+    /\ balance'.claimants = claimants
```

Comment:
From what I can find, `balance` is not a record type: https://github.com/freespek/solarkraft/blob/32f48e5f26e4cce76df4703a1b29fd1752cd1677/doc/scf24/example/timelock_mon2.tla

I know you say this spec is not valid TLA+, but having variable declarations and their types would help a lot.

Reply:
For the actual semantics of what both specs are supposed to represent, look at the [Timelock contract](https://github.com/stellar/soroban-examples/blob/v20.0.0/timelock/src/lib.rs). In particular, `balance` is an entry in the ledger with the [Balance key](https://github.com/stellar/soroban-examples/blob/186443aab0bf8b7c2673428c38708bf38cb772ab/timelock/src/lib.rs#L15), which points to the [ClaimableBalance struct](https://github.com/stellar/soroban-examples/blob/186443aab0bf8b7c2673428c38708bf38cb772ab/timelock/src/lib.rs#L34-L39) when the key is present.

Reply:
I also agree that the defs of the structs and everything should be present eventually. The idea now is to evaluate the approach as a whole, and to fill in the missing pieces later. 

Not directly related to this, but tangentially: the idea is that eventually the definitions wrt. the concrete contract could be auto-generated from the contract source code, so we won't have to write them by hand. But that's also not in the current stage, but in a subsequent one.